### ISSUE TITLE
[$2000] Background selection indicator disappears after holding down on floating menu on mWeb but not on native apps

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#18306]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open mWeb on chrome
2. Click on "New" icon
3. Hold down on an action
4. Notice the selection indicator
5. Open Android / iOS app and do the same

## Expected result:
The background selection indicator should disappear after a while just like mWeb
## Actual result:
The background selection indicator persists

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.8.8
**Reproducible in staging?:** y 
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/235791625-93e4c02b-7efa-4b0e-98e6-c07264a841fe.mp4

https://user-images.githubusercontent.com/43996225/235791705-abc2e85e-b828-46da-b94e-983c9ea4c19c.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @nathan-mulugeta
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1683019489514809

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~016a13290cdc798dc7</li>
        <li>Upwork Job ID: 1658732201930145792</li>
        <li>Last Price Increase: 2023-05-29</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by dylanexpensify** [comment]
Ah nice, yeah I agree. Do we have a current convo going in slack about this?

**Comment by dylanexpensify** [comment]
Ah nice, yeah I agree. Do we have a current convo going in slack about this?

**Comment by Beamanator** [comment]
Not that I know of, do you want to start the convo? 😅

**Comment by dylanexpensify** [comment]
Brought it up [here](https://expensify.slack.com/archives/C01SKUP7QR0/p1683194051889409)!!

**Comment by dylanexpensify** [comment]
Didn't get to this today due to prioritizing Payroll, CPM, ECUK items! Will review Monday!

**Comment by dylanexpensify** [comment]
Today:
- review convo above and summarize
- apply that to this issue

**Comment by dylanexpensify** [comment]
Melvin, I just posted above!

**Comment by dylanexpensify** [comment]
Discussion resulted in:
> So perhaps the path is:
If right-click menu, hold-press activates it
If no right-click menu, hold-press disappears/doesn’t tie to any action

Given this, this is a bug

**Comment by dylanexpensify** [comment]
@Nathan-Mulugeta can you confirm this is still active? I cannot repro on mWeb/Chrome Native android app

**Comment by Nathan-Mulugeta** [comment]
@dylanexpensify I am still able to reproduce this. I will attach a video. https://user-images.githubusercontent.com/111440031/236991784-e25c6a43-c03e-4ef7-81a3-6f315aec4019.mp4

**Comment by dylanexpensify** [comment]
Coming back from OOO! @Nathan-Mulugeta would you mind posting the video?

**Comment by twisterdotcom** [comment]
Linked to: https://github.com/Expensify/App/issues/18113

**Comment by dylanexpensify** [comment]
Updated format:
- long press on all mobile devices will have the same effect as a right click on desktop

**Comment by MonilBhavsar** [comment]
@dylanexpensify should we put this on hold for other similar issue?

**Comment by dylanexpensify** [comment]
@MonilBhavsar I think these should all be open! We came to [this](https://github.com/Expensify/App/issues/18306#issuecomment-1549275131) decision. LMK if I'm missing something!

**Comment by dylanexpensify** [comment]
[I'm heading OOO tomorrow](https://expensify.slack.com/archives/C01SKUP7QR0/p1684776773511559) for about a week so adding another BZ member while I'm away to help keep the train moving! 🚂 Thanks @mallenexpensify!!

**Comment by sobitneupane** [comment]
@dylanexpensify @Nathan-Mulugeta I don't think we have an issue here.

> Updated format:
> 
> * long press on all mobile devices will have the same effect as a right click on desktop

Long Pressing any action item (New Chat, New Group) does not trigger `onpress` event. I think it is the expected behavior.

Regarding the background color, I think it is the expected behavior and occurs throughout the app. Same is the case in other apps as well.

**Comment by twisterdotcom** [comment]
Going to take it as I have context.

**Comment by twisterdotcom** [comment]
>Long Pressing any action item (New Chat, New Group) does not trigger `onpress` event

@sobitneupane what is happening here then: https://github.com/Expensify/App/issues/18113? 

>Holding down on new icon and releasing it after a few seconds opens up the menu on Android but not on mWeb

**Comment by sobitneupane** [comment]
@twisterdotcom In https://github.com/Expensify/App/issues/18113, long pressing New icon triggers press event in native which is not the case here.

**Comment by s-alves10** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When holding down on floating menu, selection indicator doesn't disappear on native mobile apps.

### What is the root cause of that problem?

This is the default behavior of mobile native apps.

### What changes do you think we should make in order to solve the problem?

In our current project, `MenuItem` component uses `PressableWithSecondaryInteraction` internally and we can pass `onSecondaryInteraction` to `MenuItem` component to act the same as right-click. 

In cases `onSecondaryInteraction` is not provided, we need to manually check the long-press and apply styles.

We can do this in `MenuItem` component.

1. Add the state to the `MenuItem` component above this line
https://github.com/Expensify/App/blob/2f34da8fa8f5edd87f906b8fad2ec864a7e19564/src/components/MenuItem.js#L65

```
    const [pressedIn, setPressedIn] = React.useState(false);
```

2. If `onSecondaryInteraction` is not provided, we set the `pressedIn` state to false after some time(here 500ms: default delay of long press). We set this state to false in onPressOut callback as well for consistency. Add the following methods to `MenuItem` component.
```
    const onPressIn = (e) => {
        props.shouldBlockSelection && props.isSmallScreenWidth && DeviceCapabilities.canUseTouchScreen() && ControlSelection.block();
        setPressedIn(true);
        if (!props.onSecondaryInteraction) {
            setTimeout(() => {
                setPressedIn(false);
            }, 500);
        }
    }

    const onPressOut = (e) => {
        setPressedIn(false);
        ControlSelection.unblock(e);
    }
```

3. Set the props to the `PressableWithSecondaryInteraction` component.

https://github.com/Expensify/App/blob/2f34da8fa8f5edd87f906b8fad2ec864a7e19564/src/components/MenuItem.js#L104-L105

```
    onPressIn={onPressIn}
    onPressOut={onPressOut}
```

4. Update the styles.

https://github.com/Expensify/App/blob/2f34da8fa8f5edd87f906b8fad2ec864a7e19564/src/components/MenuItem.js#L109

```
                StyleUtils.getButtonBackgroundColorStyle(getButtonState(props.focused || hovered && pressedIn, pressed && pressedIn, props.success, props.disabled, props.interactive), true),
```

This works fine

<details>
    <summary>Result</summary>

https://github.com/Expensify/App/assets/126839717/994c1903-7bda-43a1-881c-a58132ac668d

</details>

### What alternative solutions did you explore? (Optional)

**Comment by twisterdotcom** [comment]
>long pressing New icon triggers press event in native which is not the case here

But our goal is the align the two. So, in this instance, isn't it more like hovering?
<img width="481" alt="image" src="https://github.com/Expensify/App/assets/9133401/0f60e849-900b-46eb-be9d-acab797e1ccc">

So, we always want it to show the highlight don't we? But we **don't** want the action to be incurred when they release because that would be a longpress, which these buttons don't have actions for.

**Comment by twisterdotcom** [comment]
@sobitneupane damn you're right.

This _also_ feels weird though doesn't it? Like, holding down your left-click and releasing it doing nothing. 🤔

Asking again: https://expensify.slack.com/archives/C01GTK53T8Q/p1685027043235089?thread_ts=1683612523.902839&cid=C01GTK53T8Q

**Comment by twisterdotcom** [comment]
Okay, no. Let's do nothing for a long click as well. That's clean.

**Comment by sobitneupane** [comment]
@twisterdotcom 

Should we update the issue description with the new findings? The issue exists on Web not on native and mWeb. So, the only change we are expecting here is to remove long press triggering action from Web. Long Pressing the menu items on New (New Chat, New Group, etc.) should do nothing.

**Comment by YousefKhawaga** [comment]
## Proposal
 [Updated](https://github.com/Expensify/App/issues/18306#issuecomment-1557268525)

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/18306#issuecomment-1561957672)

**Comment by sobitneupane** [comment]
We are still in discussion: https://expensify.slack.com/archives/C01GTK53T8Q/p1683612523902839

**Comment by MonilBhavsar** [comment]
Same ^

**Comment by twisterdotcom** [comment]
Still discussing in Slack. I think we may need to take this to a wider audience even. The core piece is whether we agree that:

1. Consistency across platforms means the same action does the same thing everywhere OR
2. Platform Consistency means adhering to what the user would expect _of that platform_.

**Comment by YousefKhawaga** [comment]
Proposal
[Updated](https://github.com/Expensify/App/issues/18306#issuecomment-1557268525)

**Comment by MonilBhavsar** [comment]
@sobitneupane could you please take a look. We arrived to [this] consensus (https://expensify.slack.com/archives/C01GTK53T8Q/p1685572307831119?thread_ts=1683612523.902839&cid=C01GTK53T8Q) on slack discussion

**Comment by sobitneupane** [comment]
@MonilBhavsar Is [this](https://expensify.slack.com/archives/C01GTK53T8Q/p1685605241472309?thread_ts=1683612523.902839&cid=C01GTK53T8Q) the required behavior?  If so, I don't think any change is required here. We already have [the behavior](https://expensify.slack.com/archives/C01GTK53T8Q/p1685605241472309?thread_ts=1683612523.902839&cid=C01GTK53T8Q).

**Comment by twisterdotcom** [comment]
Well we decided what would happen when you unpressed, but I think that when you hold the press down, the highlight should remain, which is what @WikusKriek is fixing.

**Comment by sobitneupane** [comment]
@WikusKriek 
> What is the root cause of that problem?
On mobile platforms the "pressed" event stays true on a long press unlike mWeb and desktop. This is why the ButtonBackgroundColorStyle is still active while holding in the menu item on mobile.

IMO the root cause here is: the default behavior of mWeb is to allow text selection on long press. When long pressed on mWeb, text is selected and background color changes back to normal. This is the case for almost all menu items in our app.

If we block the text selection, the highlighted background color will remain as long as menu item is pressed. But the question is if we want to do so? Might require some discussion.

@WikusKriek I am not sure how your solution will work. Will the background be highlighted and text selected?

**Comment by twisterdotcom** [comment]
Just want to interject and say I really _really_ appreciate all the back and forth here, especially from you @sobitneupane and @WikusKriek. I never knew how confusing such a seemingly simple thing could be to get to the bottom of.

Agree, I'd love to see a screen recording. And this, to my untrained eye, sounds good:
>I am not blocking text selection, my solution only changes whether or not the ButtonBackgroundStyle is active. I have an additional flag that is controlled by the long-press. And with this new flag and the previous "pressed" flag I control the styling(pressed || longPress to activate background style).

**Comment by sobitneupane** [comment]
@WikusKriek Can we see how it look on mWeb. I don't think there was any issue on native.

**Comment by sobitneupane** [comment]
So, going back to [this](https://github.com/Expensify/App/issues/18306#issuecomment-1580553120), do we want to block text selection on long press in mWeb for consistency across all platforms or do nothing as it is expected in mWeb for text to be selected on long press?

cc: @twisterdotcom @MonilBhavsar

**Comment by twisterdotcom** [comment]
>do nothing as it is expected in mWeb for text to be selected on long press?

This would be my preference.

**Comment by twisterdotcom** [comment]
Are we moving on this one now?

**Comment by sobitneupane** [comment]
@twisterdotcom I think [this](https://github.com/Expensify/App/issues/18306#issuecomment-1581809188) was the only issue here. I don't think we have any thing to fix.

cc: @WikusKriek

**Comment by twisterdotcom** [comment]
Whoah! Sorry for missing these updates. Okay, so we should close right @sobitneupane? I'd sort of be happy to pay out an equal amount to a bug reporting bonus for for @WikusKriek and @sobitneupane, even if we aren't going to go ahead with any change given the work you've put into getting us here. Does that sound okay?

**Comment by sobitneupane** [comment]
Yup @twisterdotcom. We are good to close this.

**Comment by twisterdotcom** [comment]
Two offers of $250 sent. I'll close this out now and just get those paid whenever you accept. Thanks all!
